python-client/giskard/models/model_explanation.py

```diff
@@ -6,46 +6,114 @@
+    df = model.prepare_dataframe(df, column_dtypes=dataset.column_dtypes, target=dataset.target)
+
+    if dataset.target in df.columns:
+        prepared_dataset = Dataset(df, column_types=dataset.column_types, target=dataset.target)
+    else:
+        prepared_dataset = Dataset(df, column_types=dataset.column_types)
+
+    # Make sure column order is the same as in the dataset.df.
+    columns_original_order = (
+        model.meta.feature_names
+        if model.meta.feature_names
+        else [c for c in dataset.df.columns if c in prepared_dataset.df.columns]
+    )
+
+    prepared_df = prepared_dataset.df[columns_original_order]
+    return prepared_df
+
+
+def background_example(df: pd.DataFrame, input_types: Dict[str, str]) -> pd.DataFrame:
+    """Create back-ground example for the SHAP explainer as a mode/median of features of the data to explain."""
+    median = df.median()
+    example = df.mode(dropna=False).head(1)
+    num_columns = [key for key in list(df.columns) if input_types.get(key) == "numeric"]
+
+    # Replace a numerical features' mode on their median.
+    for column in num_columns:
+        example[column] = median[column]
+
+    example = example.astype(df.dtypes)
+    return example
 
-    input_df = prepare_df(pd.DataFrame([input_data]))
+
+def explain_full(model: BaseModel, dataset: Dataset) -> np.ndarray:
+    """Perform SHAP values calculation for samples of a given dataset."""
+    # Prepare background sample to be used in the KernelSHAP.
+    background_df = model.prepare_dataframe(dataset.df, dataset.column_dtypes, dataset.target)
+    background_sample = background_example(background_df, dataset.column_types)
+
+    # Prepare input data for explanation.
+    input_df = prepare_df(dataset.df, model=model, dataset=dataset)
+
+    # Obtain SHAP explanations.
+    explainer = KernelExplainer(model.predict_df, background_sample, feature_names=input_df.columns, keep_index=True)
+    shap_values = explainer.shap_values(input_df, silent=True)
+    return shap_values
+
+
+def explain_with_shap(model: BaseModel, dataset: Dataset) -> ShapResult:
+    """Get SHAP explanation result."""
+    # Obtain SHAP explanations.
+    shap_values = explain_full(model, dataset)
+
+    # For classification, take SHAP of prediction with the highest probability.
+    if model.is_classification:
+        shap_values = _get_cls_prediction_explanation(model, dataset, shap_values)
+
+    # Put shap and feature names to the Explanation object for a convenience.
+    feature_names = model.meta.feature_names or list(dataset.df.columns.drop(dataset.target, errors="ignore"))
+    feature_types = {key: dataset.column_types[key] for key in feature_names}
+    shap_explanations = Explanation(values=shap_values, data=dataset.df[feature_names], feature_names=feature_names)
+    return ShapResult(shap_explanations, feature_types, feature_names)
+
+
+def explain_one(model: BaseModel, dataset: Dataset, input_data: Dict) -> np.ndarray:
```

Comment:
let's refactor this into: `_calculate_sample_shap_values`

Reply:
Done